### ISSUE TITLE
Smallest mailbox router pool occasionally eating 100% CPU

### ISSUE THREAD

**Issue by cmbaxter** [akka/akka#19949]
I originally posted this on the Akka User groups page here:

https://groups.google.com/forum/#!topic/akka-user/2vA2npqq2aU

It appears that sometimes the smallest mailbox routing logic can get stuck in a do/while loop in `AbstractNodeQueue.peekNode`.  This is sporadic in nature, but if I run enough volume through my server I can eventually get it to do this.  I did not see this in the 2.3.x series, and only noticed it after we upgraded to 2.4.2.  The stack of the stuck thread looks like this:

> "AqutoServices-akka.actor.default-dispatcher-17" #32 prio=5 os_prio=0 tid=0x00007f404c016000 nid=0x37d1 runnable [0x00007f406445b000]
   java.lang.Thread.State: RUNNABLE
	at akka.dispatch.AbstractNodeQueue.peekNode(AbstractNodeQueue.java:42)
	at akka.dispatch.AbstractNodeQueue.count(AbstractNodeQueue.java:70)
	at akka.dispatch.NodeMessageQueue.numberOfMessages(Mailbox.scala:376)
	at akka.dispatch.Mailbox.numberOfMessages(Mailbox.scala:102)
	at akka.actor.dungeon.Dispatch$class.numberOfMessages(Dispatch.scala:34)
	at akka.actor.ActorCell.numberOfMessages(ActorCell.scala:374)
	at akka.routing.SmallestMailboxRoutingLogic.numberOfMessages(SmallestMailbox.scala:132)
	at akka.routing.SmallestMailboxRoutingLogic.selectNext(SmallestMailbox.scala:67)
	at akka.routing.SmallestMailboxRoutingLogic.select(SmallestMailbox.scala:37)
	at akka.routing.Router.route(Router.scala:116)

I'm going to try and switch those actors over to a round robin router pool for now.  I would appreciate any insight you guys might have on this.  Viktor seems to think that it might have to do with: https://github.com/akka/akka/issues/19216

**Comment by rkuhn** [comment]
Thanks for the report, this was indeed overlooked when fixing the value nepotism issue: `count` can well see the `.next` reference being nulled out as per the previous patch, which will then not change (because it was nulled while the code is written with the assumption that the `null` was spurious).

The solution is that `count` cannot use `peekNode`.

**Comment by cmbaxter** [comment]
Okay.  I'll keep an eye on this issue to see what the fix looks like.  In the mean time, the switch away from using smallest mailbox for the very few actors that we had using it seems to have avoided this issue.

**Comment by MQ-EL** [comment]
I think there is an another problem may not be observered.

```scala
      public final int count() {
        int count = 0;
        // if n.next() retrun null, this n may be not the real header node. 
        // imagine the pullNode catch up with it, and set this n.next to null.
        for(Node<T> n = peekNode();n != null && count < Integer.MAX_VALUE; n = n.next())
          ++count;
        return count;
    }
```

so, it also has a problem(not stuck but wrong number returned) in ```count``` even if without to use ```peekNode```.

**Comment by rkuhn** [comment]
Since using `count` is racy no matter which side performs it, there is no “wrong number”; technically it would be correct to return zero in all cases, everything else is best effort.

**Comment by MQ-EL** [comment]
"wrong number" will be happened when something happened by order as follows :

assumption: the ABNQ object contains 1000 nodes.

1. in **thread 1**, call ```count```: ```peekNode``` wins, loops 10 cycle, up to 11st node and the count now is 10
2. in **thread 2**, consumer (actor) ```pollNode``` from ABNQ, and catch up to 11st node, set 11st node's next to null.
3. in **thread 1**, get the 11st's next, find it is null, then quit the loop, return 10, which is "wrong number".

In either case, the "correct number" should not be 10. 990 or 1000 is right

**Comment by MQ-EL** [comment]
The ```count``` is an O(n) operation in many unbound Queues include ABNQ. Avoid use it when need to compute size in production environment as far as possible. Just my two cents :)

**Comment by MQ-EL** [comment]
Execute me, @rkuhn , may i ask a stupid question: why the triaged tag was removed? Thanks In Advance

**Comment by patriknw** [comment]
no worries, we are filtering tickets at the moment, it will be put back when we have re-triaged them

**Comment by MQ-EL** [comment]
ok, Thanks patriknw

**Comment by patriknw** [comment]
this has been fixed in master (2.4.3) and release-2.3 (2.3.15)
